driver-core/src/main/com/mongodb/internal/TimeoutContext.java

```diff
@@ -0,0 +1,371 @@
+            return false;
+        }
+        long remaining = timeout.remaining(MILLISECONDS);
+        return remaining <= 0 || minRoundTripTimeMS > remaining;
+    }
+
+    /**
+     * Returns the remaining {@code timeoutMS} if set or the {@code alternativeTimeoutMS}.
+     *
+     * @param alternativeTimeoutMS the alternative timeout.
+     * @return timeout to use.
+     */
+    public long timeoutOrAlternative(final long alternativeTimeoutMS) {
+        Long timeoutMS = timeoutSettings.getTimeoutMS();
+        if (timeoutMS == null) {
+            return alternativeTimeoutMS;
+        }
+        return timeoutRemainingMS();
+    }
+
+    /**
+     * Calculates the minimum timeout value between two possible timeouts.
+     *
+     * @param alternativeTimeoutMS the alternative timeout
+     * @return the minimum value to use.
+     */
+    public long calculateMin(final long alternativeTimeoutMS) {
+        Long timeoutMS = timeoutSettings.getTimeoutMS();
+        if (timeoutMS == null) {
+            return alternativeTimeoutMS;
+        } else if (timeoutMS == 0) {
+            return alternativeTimeoutMS;
+        } else if (alternativeTimeoutMS == 0) {
+            return timeoutRemainingMS();
+        } else {
+            return Math.min(timeoutRemainingMS(), alternativeTimeoutMS);
+        }
+    }
+
+    public TimeoutSettings getTimeoutSettings() {
+        return timeoutSettings;
+    }
+
+    public long getMaxAwaitTimeMS() {
+        return timeoutSettings.getMaxAwaitTimeMS();
+    }
+
+    public long getMaxTimeMS() {
+        return notNull("Should never be null", maxTimeSupplier.get());
+    }
+
+    private long calculateMaxTimeMS() {
+        long maxTimeMS = timeoutOrAlternative(timeoutSettings.getMaxTimeMS());
+        if (timeout == null || timeout.isInfinite()) {
+            return maxTimeMS;
+        }
+        if (minRoundTripTimeMS >= maxTimeMS) {
+            throw createMongoTimeoutException();
+        }
+        return maxTimeMS - minRoundTripTimeMS;
+    }
+
+    public void setMaxTimeSupplier(final MaxTimeSupplier maxTimeSupplier) {
+        this.maxTimeSupplier = maxTimeSupplier;
+    }
+
+    public void resetToDefaultMaxTimeSupplier() {
+        this.maxTimeSupplier = this::calculateMaxTimeMS;
+    }
+
+    public Long getMaxCommitTimeMS() {
+        Long maxCommitTimeMS = timeoutSettings.getMaxCommitTimeMS();
+        return timeoutOrAlternative(maxCommitTimeMS != null ? maxCommitTimeMS : 0);
+    }
+
+    public long getReadTimeoutMS() {
+        return timeoutOrAlternative(timeoutSettings.getReadTimeoutMS());
+    }
+
+    public long getWriteTimeoutMS() {
+        return timeoutOrAlternative(0);
+    }
+
+    public int getConnectTimeoutMs() {
+        return (int) calculateMin(getTimeoutSettings().getConnectTimeoutMS());
+    }
+
+    public void resetTimeout() {
+        assertNotNull(timeout);
+        timeout = calculateTimeout(timeoutSettings.getTimeoutMS());
+    }
```

Comment:
Consider changing to `resetTimeoutIfPresent`, with an if check on null rather than the assert. All prod usages perform this check before calling.

Reply:
Done in https://github.com/mongodb/mongo-java-driver/pull/1449.